src/H5Dio.c

```diff
@@ -107,6 +107,17 @@ H5D__read(size_t count, H5D_dset_io_info_t *dset_info)
 
     FUNC_ENTER_NOAPI(FAIL)
 
+#ifdef H5_HAVE_PARALLEL
+    /* Reset the actual io mode properties to the default values in case
+     * the DXPL (if it's non-default) was previously used in a collective
+     * I/O operation.
+     */
+    if (!H5CX_is_def_dxpl()) {
```

Comment:
Here I needed to move the setting of the optimization method higher up so if something fails in this function the returned value is sane.

Reply:
Maybe put this in H5D__ioinfo_init() instead?

Reply:
Thinking about this more, I feel like this may even belong higher up in the library since `H5D__read` is called in a few other places. It also seems like good compartmentalization to just have `H5D__ioinfo_init` init the `io_info` structure without doing much else.